pkg/util/requestbuffers.go

```diff
@@ -3,19 +3,51 @@ package util
 
 import (
 	"bytes"
-)
+	"sync"
 
-// The maximum buffer size allowed in the pool.
-// A sane default value of 1MB is chosen, which should be sufficient for most use cases, though it could be
-// revisited if necessary.
-const maxInPoolRequestBufferSize = 1024 * 1024
+	"github.com/prometheus/prometheus/util/pool"
+)
 
-// Pool is an abstraction of sync.Pool, for testability.
+// Pool is an abstraction for a pool of byte slices.
 type Pool interface {
-	// Get a pooled object.
-	Get() any
-	// Put back an object into the pool.
-	Put(any)
+	// Get returns a new byte slices that fits the given size.
+	Get(sz int) []byte
+
+	// Put puts a slice back into the pool.
+	Put(s []byte)
+}
+
+type bufferPool struct {
+	p sync.Pool
+}
+
+func (p *bufferPool) Get(_ int) []byte { return p.p.Get().([]byte) }
+func (p *bufferPool) Put(s []byte)     { p.p.Put(s) } //nolint:staticcheck
+
+// NewBufferPool returns a new Pool for byte slices.
+func NewBufferPool() Pool {
+	return &bufferPool{
+		p: sync.Pool{
+			New: func() interface{} {
+				return make([]byte, 0, 256*1024)
+			},
+		},
+	}
+}
+
+type bucketedBufferPool struct{ p *pool.Pool }
+
+func (p *bucketedBufferPool) Get(sz int) []byte { return p.p.Get(sz).([]byte) }
+func (p *bucketedBufferPool) Put(s []byte)      { p.p.Put(s) } //nolint:staticcheck
+
+// NewBucketedBufferPool returns a new Pool for byte slices with bucketing.
+// The pool will have buckets for sizes from minSize to maxSize increasing by the given factor.
+func NewBucketedBufferPool(minSize, maxSize int, factor float64) Pool {
```

Comment:
Is this somehow different from the bucketed pool we already have?

https://github.com/grafana/mimir/blob/45ccdb0057aa53d97f9595105c4e1ffd288680c3/pkg/util/pool/bucketed_pool.go#L24

Reply:
Didn't know about this one. :)

Let me see...

Reply:
_changed @ 0d6f5afe5a2e1283679e56cd517273b4c27b7a4e_